libass/ass_alloc.c

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (C) 2020 libass contributors
+ *
+ * This file is part of libass.
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include "config.h"
+#include "ass_compat.h"
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "ass_alloc.h"
+#include "ass_utils.h"
+
+// strdup and strndup are only part of ISO C since C2x; libass only requires C99
+#ifndef HAVE_STRDUP
```

Comment:
Right, they should use regular `malloc` to avoid duplicate messages on failure, thanks.
But I'd prefer to keep the fallback implementation conditional, as libc's implementation if available is potentially better optimised.